File: Capítulo_11/Heap Binário/MinHeap/Extraindo os valores mínimo e máximo do heap/Código em Python/extract.py
# Compare elements
Compare = {
    "LESS_THAN": -1,
    "BIGGER_THAN": 1
}

# Função default_compare
def default_compare(a, b):
    if a == b:
        return 0
    return Compare["LESS_THAN"] if a < b else Compare["BIGGER_THAN"]

# Função reverse_compare
def reverse_compare(compare_fn):
    return lambda a, b: compare_fn(b, a)

# Função swap´
def swap(array, a, b):
    temp = array[a]
    array[a] = array[b]
    array[b] = temp

# Classe MinHeap
class MinHeap:
    def __init__(self, compare_fn = default_compare):
        self.compare_fn = compare_fn
        self.heap = []

    # Método extract(): esse método remove o value mínimo (heap mínimo) ou máximo (heap máximo) e devolve esse valor.
    def extract(self):
        if self.is_empty():
            return None
        if self.size() == 1:
            return self.get_heap().pop(0)
        removed_value = self.get_heap()[0]
        heap = self.get_heap()
        heap[0] = heap.pop()
        self.sift_down(0)
        return removed_value

File: Capítulo_11/Heap Binário/MinHeap/Extraindo os valores mínimo e máximo do heap/extract.js
import { Compare, defaultCompare, swap } from "../../../models/util";

// Classe MinHeap
export class MinHeap {
    constructor(compareFn = defaultCompare) {
        this.compareFn = compareFn;
        this.heap = [];
    }

    // Método extract(): esse método remove o value mínimo (heap mínimo) ou máximo (heap máximo) e devolve esse valor.
    extract() {
        if (this.isEmpty()) {
            return undefined;
        }
        if (this.size() === 1) {
            return this.heap.shift();
        }
        const removedValue = this.heap.shift();
        this.siftDown(0);
        return removedValue;
    }
}
